pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/table/JSymbolTable.java

```diff
@@ -0,0 +1,121 @@
+ * shortcomings of the current symbol table make the current typeres duplicate
+ * logic, and ultimately perform tasks that are not its responsibility,
+ * which is probably why {@link ClassTypeResolver} is so huge and nasty.
+ *
+ * <p>Having an abstraction layer to unify them allows the AST analyses to
+ * be complementary, and rely on each other, instead of being so self-reliant.
+ * The abstraction provided by {@link JElementSymbol} may in the future be used
+ * to build global indices of analysed projects to implement multifile analysis.
+ *
+ * <p>The goals of this rewrite should be:
+ * <ul>
+ *      <li>To make our language analyses passes share information and be complementary
+ *      <li>To have a symbol table that is precise and exhaustive enough that all rules can depend on it
+ *      <li>To modularize the system so that it's more easily testable and documentable
+ * </ul>
+ *
+ * @author Clément Fournier
+ * @since 7.0.0
+ */
+// @formatter:on
+@Experimental
+public interface JSymbolTable {
+
+    /**
+     * Returns the parent of this table, that is, the symbol table that will be
+     * delegated to if this table doesn't find a declaration.
+     *
+     * @return a symbol table, or null if this is the top-level symbol table
+     */
+    JSymbolTable getParent();
+
+
+    /**
+     * Resolves the type referred to by the given name. This must be a simple name,
+     * ie, parameterized types and array types are not available. Primitive types are
+     * also not considered because it's probably not useful.
+     *
+     * <p>The returned type reference may be a {@link JResolvableClassSymbol}, a
+     * {@link JClassSymbol} if the type was already resolved, or a {@link JTypeParameterSymbol}.
+     *
+     * <p>Doesn't handle primitive types (there's no symbol for them).
+     *
+     * @param simpleName Simple name of the type to look for
+     *
+     * @return The type reference if it can be found, otherwise an empty optional
+     */
+    Optional<? extends JSimpleTypeSymbol> resolveTypeName(String simpleName);
+
+
+    /**
+     * Finds the variable or field to which the given simple name refers in the scope of
+     * this symbol table.
+     *
+     * @param simpleName simple name of the value to find
+     *
+     * @return The reference to the variable if it can be found, otherwise an empty optional
+     */
+    Optional<JValueSymbol> resolveValueName(String simpleName);
+
+
+    /**
+     * Finds all accessible methods that can be called with the given simple name
+     * on an implicit receiver in the scope of this symbol table. The returned methods may
+     * have different arity and parameter types.
+     *
+     * @param simpleName Simple name of the method
+     *
+     * @return A stream yielding all methods with the given name accessible and applicable to the
+     * implicit receiver in the scope of this symbol table.
+     */
+    Stream<JMethodSymbol> resolveMethodName(String simpleName);
```

Comment:
This is the only place where we use a stream instead of a list, any reason for that?

Reply:
Possibly, looking up a method may involve looking up all the supertypes of the class in scope (and enclosing classes and their supertypes, for inner/anonymous classes). So since this might be costly, doing it lazily and stopping the hierarchy exploration as soon as we find a matching method might improve performance. A return type that is not a list also documents that the methods may not actually be prefetched.

But maybe all of this should be mentioned on the doc of the method.

Reply:
The thing is, unless the method I'm looking for has 0 args / I have an **EXACT** match on **all** parameters, I'll always need all to determine which overload best fits my arguments and that requires the complete list.

Probably most methods will fit that bill 'though…

Reply:
May be... Tbh it's too soon to say, we'll have to profile this for different scenarios when it's implemented and used by type resolution.